goalc/compiler/compilation/Asm.cpp

```diff
@@ -507,125 +775,178 @@ Val* Compiler::compile_asm_abs_vf(const goos::Object& form, const goos::Object&
+      {{"color", {false, goos::ObjectType::SYMBOL}}, {"ftf", {true, goos::ObjectType::INTEGER}}});
   bool color = true;
   if (args.has_named("color")) {
     color = get_true_or_false(form, args.named.at("color"));
   }
 
   auto dest = compile_error_guard(args.unnamed.at(0), env)->to_reg(env);
-  if (!dest->settable() || dest->ireg().reg_class != RegClass::VECTOR_FLOAT) {
-    throw_compiler_error(
-        form, "Invalid destination register for a vector float 3-arg math form. Got a {}.",
-        dest->print());
-  }
+  auto src = compile_error_guard(args.unnamed.at(1), env)->to_reg(env);
+  check_vector_float_regs(form, env, {{"destination", dest}, {"source", src}});
 
-  auto src1 = compile_error_guard(args.unnamed.at(1), env)->to_reg(env);
-  if (src1->ireg().reg_class != RegClass::VECTOR_FLOAT) {
-    throw_compiler_error(
-        form, "Invalid first source register for a vector float 3-arg math form. Got a {}.",
-        src1->print());
+  u8 ftf = args.named.at("ftf").as_int();
+  if (ftf > 3) {
+    throw_compiler_error(form, "The value {} is out of range for ftf (0-3 inclusive).", ftf);
   }
 
-  auto src2 = compile_error_guard(args.unnamed.at(2), env)->to_reg(env);
-  if (src2->ireg().reg_class != RegClass::VECTOR_FLOAT) {
-    throw_compiler_error(
-        form, "Invalid second source register for a vector float 3-arg math form. Got a {}.",
-        src2->print());
-  }
+  // VSQRT in the VU stores its result in a single 32bit `Q` Register, it does not compute the
+  // packed division result
+  //
+  // Because of this, we need to blend the relevent component into a cleared register and then
+  // perform the SQRT
+  //
+  // Why do we even bother using VSQRTPS instead of FSQRT? Because otherwise in x86, you have to use
+  // the FPU stack Registers are nicer.
 
-  u8 mask = 0b1111;
-  if (args.has_named("mask")) {
-    mask = args.named.at("mask").as_int();
-    if (mask > 15) {
-      throw_compiler_error(form, "The value {} is out of range for a blend mask (0-15 inclusive).",
-                           mask);
-    }
-  }
+  // Splat src's value into the dest reg, keep it simple, this way no matter which vector component
+  // is accessed from the final result will be the correct answer
+  auto temp_reg = env->make_vfr(dest->type());
+  env->emit_ir<IR_SplatVF>(color, temp_reg, src, ftf_fsf_to_vector_element(ftf));
 
-  // If there is a broadcast register, splat that float across the entire src2 register before
-  // performing the operation For example vaddx.xyzw vf10, vf20, vf30
-  // vf10[x] = vf20[x] + vf30[x]
-  // vf10[y] = vf20[y] + vf30[x]
-  // vf10[z] = vf20[z] + vf30[x]
-  // vf10[w] = vf20[w] + vf30[x]
-  if (broadcastElement != emitter::Register::VF_ELEMENT::NONE) {
-    auto temp_reg = env->make_vfr(dest->type());
-    env->emit_ir<IR_SplatVF>(color, temp_reg, src2, broadcastElement);
+  // Perform the operation, you can't seem to VSQRTPS with the same register, which is why we needed
```

Comment:
Hmmm. I think the encoding of `vsqrtps` may be slightly wrong and causing this issue.  Can we try this without the temp register with the encoding fixed?

Reply:
Hmm, good catch (with below info) however, I'm not able to get things to work.  I get Illegal Instructions, I'll commit what I have